velox/expression/PrestoCastHooks.cpp

```diff
@@ -65,6 +70,109 @@ Expected<int32_t> PrestoCastHooks::castStringToDate(
   return util::castFromDateString(dateString, util::ParseMode::kStandardCast);
 }
 
+namespace {
+
+// Returns an user error if 'v' is the equivalent to 'valid' but has different
+// letter cases. Returns OK if 'v' is exactly the same as 'valid' or 'v' is a
+// completely different string.
+template <TypeKind ToKind>
+Status checkInvalidCase(const StringView& v, const char* valid, size_t length) {
+  if (v.size() != length) {
+    return Status::OK();
+  }
+  bool invalidCase = false;
+  for (auto i = 0; i < length; ++i) {
+    if (v.data()[i] == valid[i]) {
+      continue;
+    }
+    if (std::tolower(v.data()[i]) == std::tolower(valid[i])) {
+      invalidCase = true;
+    } else {
+      // Not a case-insensitive stirng of 'valid'.
+      return Status::OK();
+    }
+  }
+
+  if (invalidCase) {
+    return Status::UserError(
+        "Cannot cast {} to {}", v, TypeTraits<ToKind>::name);
+  }
+  return Status::OK();
+}
+
+// Returns a user error if 'v' is equivalent to 'invalid' regardless of the
+// letter case.
+template <TypeKind ToKind>
+Status
+RejectInvalidString(const StringView& v, const char* invalid, size_t length) {
+  if (v.size() != length) {
+    return Status::OK();
+  }
+  for (auto i = 0; i < length; ++i) {
+    if (std::tolower(v.data()[i]) != std::tolower(invalid[i])) {
+      // Not a case-insensitive stirng of 'invalid'.
+      return Status::OK();
+    }
+  }
+  return Status::UserError("Cannot cast {} to {}", v, TypeTraits<ToKind>::name);
+}
+
+// Allows "Infinity" and "-Infinity" but not equivalent strings with different
+// letter cases. Also not allows "Inf", "-Inf", and their equivalent strings.
+// Strings not equivalent to these four are allowed since casting from a number
+// exceeding the range of the floating-point type can result in an infinity.
+template <TypeKind ToKind>
+Status checkInvalidInfinity(const StringView& v) {
+  Status result = checkInvalidCase<ToKind>(v, "Infinity", 8);
```

Comment:
This is a lot of checking. Wondering what's the impact on performance? Would it be possible to benchmark casting invalid strings that go through all these checks? Is there a way to implement this a bit more efficiently?